#  Assignment 5

This assignment covers material on matrices, norms, condition numbers and eigensystems.

The breakdown of the marks is as follows:

Question 0 - [1 mark]  
Question 1 - [7 marks]  
Questino 2 - [6 marks]  
Question 3 - [7 marks]  
Question 4 - [9 marks]  

# Question 0

You should have saved this notebook to your ACM20030-Assignments folder. Now:
- commit the notebook to your local repository
- push the changes to the GitHub

In [1]:
import numpy as np
from numpy import linalg as la
np.set_printoptions(linewidth=200)  # This stops line wrapping of larger matrices

# Question 1 - practice manipulating arrays

This questions asks you write a few functions for manipulating arrays of data

## Q1 a

Write a function that checks if a given $n\times n$ matrix is symmetric. Use only loops and `.shape` function from numpy (do not use any other special numpy functions).

In [2]:
def symmetric(a):
    
    (r,c) = a.shape #r = row, c = collumn
    # Must be square to be symmetric
    if (r != c):
        print('Matrix must be square')
        return 0
    
    #Compare a[i,j] to a[j,i], return 0 if any instance is false (Output of 0 can be used later if needed)
    i = 0
    while i < r:
        j = 0
        while j < c:
            if a[i,j] != a[j,i]:
                print ('Matrix is not symmetric')
                return 0
            j += 1
        i += 1

    #No instance was false, therefore matrix is symmetric
    print('Matrix is Symmetric')
    return 1

Test your function on the two matrices below

In [3]:
A = np.random.rand(5,5)
SymMatrix = (A + A.T)/2

In [4]:
symmetric(A);

Matrix is not symmetric


In [5]:
symmetric(SymMatrix);

Matrix is Symmetric


## Q1 b

Write a function to transpose a general $n\times n$ matrix. Do this by writing loops and not using any special functions like `np.transpose()`, `A.T`, or any functions from any other library (`np.zeroes`, `np.ones`, `np.shape`, etc are all ok to use).

In [6]:
def transpose(a):

    (r,c) = a.shape
    
    #Create new matrix b and write a[i,j] into b[j,i]
    b = np.zeros((r,c))
    i = 0
    while i < r:
        j = 0
        while j < c:
            b[j,i] = a[i,j]
            j += 1
        i += 1
    return b

Check your code by transposing a matrix by checking that $A^T \neq A$ and $(A^T)^T - A = 0$

In [7]:
#Visual Test
A = np.random.rand(2,2);
print(A)
print('\n',transpose(A))

[[0.12512921 0.88304593]
 [0.06504489 0.01639706]]

 [[0.12512921 0.06504489]
 [0.88304593 0.01639706]]


In [8]:
A = np.random.rand(6,6);
AT = transpose(A)
ATT = transpose(AT)

#Test if AT != A
(r,c) = A.shape
i = 0
while i < r:
    j = 0
    while j < c:
        if A[i,j] != AT[i,j]:
            Condition_A = 1
            break
        else:
            Condition_A = 0
            break
        j += 1
    i += 1

#Test (AT)T - A = 0
i = 0
while i < r:
    j = 0
    while j < c:
        if ATT[i,j] - A[i,j] == 0:
            Condition_B = 1
            break
        else:
            Condition_B = 0
            break
        j += 1
    i += 1

#Test Condition_A and Condition_B simultaneously
if (Condition_A and Condition_B):
    print('A has been correctly transposed')
else:
    print('A has Not been correctly transposed')

A has been correctly transposed


## Q1c

Write a function that finds the maximum entry in a matrix (where the matrix does not need to be square). Do this by writing loops and not using any special functions like `np.max` or any functions from any other library.

In [9]:
def maxentry(a):
    (r,c) = a.shape
    i = 0
    maxentry = 0
    while i < r:
        j = 0
        while j < c:
            if a[i,j] > maxentry:
                maxentry = a[i,j]
            j += 1
        i += 1
    return maxentry

Check your result by comparing to `np.max(A)` for the below random matrix

In [10]:
A = np.random.rand(4,5)

print('maxentry function:', maxentry(A))
print('np.max() function:', np.max(A))

maxentry function: 0.9374053403905801
np.max() function: 0.9374053403905801


Commit your notebook to git and push it to GitHub.

# Question 2:  matrix norms

In the lectures we focused on the $L^2$-norm. We also briefly mentioned the $L^1$ and $L^\infty$ norm. Let's look at calculating those as well for a general $m\times n$ matrix.

# Q2a

The $L^1$ norm is given by

$$ \lVert A \rVert_1 = \max_j \sum^m_{i=1} |a_{ij}| $$

i.e., it is the max of the sums of the absolute values of each column. Write a function that computes the $L^1$-norm using only loops. The function should use no special numpy functions except `np.abs()` (`np.shape`, `np.inf`, etc are ok to use).

In [11]:
def L1norm(a):
    (r,c) = a.shape
    Csum = 0
    Csummax = 0
    
    j = 0
    while j < c:
        i = 0
        while i < r:
            Csum += abs(a[i,j])
            i += 1
        if Csum > Csummax:
            Csummax = Csum
        #Re-set Csum to 0
        Csum = 0
        j += 1
        
    return Csummax

Check your result against NumPy's `la.norm(A,1)` function for the following random matrix

In [12]:
A = np.random.rand(3,4)

print('L1norm function:', L1norm(A))
print('np.norm function:', la.norm(A, 1))

L1norm function: 1.9977715504469207
np.norm function: 1.9977715504469207


## Q2b

The $L^\infty$ norm is given by 

$$\lVert A \rVert_\infty = \max_i \sum_{j=1}^n |a_{ij}|$$

i.e., it is the max of the sums of the absolute values of each row. Write a function that computes the $L^\infty$-norm using only loops and no special numpy functions.

In [13]:
def Linfnorm(a):
    (r,c) = a.shape
    Rsum = 0
    Rsummax = 0
    
    i = 0
    while i < r:
        j = 0
        while j < c:
            Rsum += abs(a[i,j])
            j += 1
        if Rsum > Rsummax:
            Rsummax = Rsum
        #Re-set Csum to 0
        Rsum = 0
        i += 1
        
    return Rsummax

Check your result against NumPy's `la.norm(A,np.inf)` function for the following random matrix

In [14]:
A = np.random.rand(3,4)

print('Linfnorm function:', Linfnorm(A))
print('np.norm function:', la.norm(A, np.inf))

Linfnorm function: 2.3319136685808814
np.norm function: 2.3319136685808814


Commit your notebook to git and push it to GitHub.

# Question 3: ill-conditions systems of equations

In this question you can use any of NumPy's functions.

## Q3a

Which of the following three matrices, $A1, A2, A3$, are ill-conditioned. State the reason for your conclusion.

In [15]:
A1 = np.array([[ 2.852416473483e-01,  3.803483118694e-01,  2.697704226632e-01,  6.810611849029e-01],
       [ 6.622827828376e-01,  6.447943415836e-01,  4.387842058829e-01,  1.530575702376e+00],
       [ 3.672130860703e-01,  8.309522079704e-01,  9.975487474007e-01,  5.678296337101e-01],
       [ 4.036490000599e-01,  6.865978225967e-06,  8.788095913773e-01, -7.150472427928e-02]])

A2 = np.array([[0.988746460106, 0.351486109477, 0.024911477532, 0.384039749683],
       [0.282104657306, 0.239884151775, 0.36170151991 , 0.903082266861],
       [0.59333290915 , 0.637178863968, 0.241229214635, 0.685595037901],
       [0.013591747828, 0.917926970364, 0.458689093885, 0.157821121254]])

A3 = np.array([[0.131923621844, 0.245316154195, 0.732362341613, 0.957047891764],
       [0.91290446184 , 0.512312007959, 0.391938831444, 0.092509684772],
       [0.648708523652, 0.608576675797, 0.576773587276, 0.065385122929],
       [0.502955480393, 0.084082960221, 0.659487390444, 0.662958739412]])

In [16]:
#Find the condition numbers (Larger ones are ill-conditioned)

print("Condition number of A1:\t", la.cond(A1,2))
print("Condition number of A2:\t", la.cond(A2,2))
print("Condition number of A3:\t", la.cond(A3,2))

Condition number of A1:	 13334936994.070848
Condition number of A2:	 21.552466488381846
Condition number of A3:	 15.12244969941062


### Matrix A1 has a very large conditition number, hence we can say that A1 is close to singular
This means that A1 is most likely ill-conditioned (without explicitly proving it)

# Q3b

With $b = [1,2,3,4]$ and $\delta b = [0.1,0,0,0]$ for $A1$ and $A2$ calculate the percentage change in $x$ due to a small change in $b$, i.e., calculate 

$$100\times\frac{\lVert \delta x_i\rVert_2}{\lVert x_i \rVert_2}\quad\text{and}\quad 100\times\frac{\lVert \delta b\rVert_2}{\lVert b \rVert_2}$$

where $(x_i+\delta x_i)  = A_i^{-1}(b+\delta b)$ and $i\in\{1,2\}$.

In [17]:
def GaussianElimination(A, b):
    n = A.shape[1]
    
    # Append the vector b as a column to the matrix A
    A1 = np.c_[A,b]
    
    i = 0
    while i < n - 1:
        j = i+1
        while j < n:
            A1[j, i+1:] = A1[j, i+1:] - A1[i, i+1:]*A1[j,i]/A1[i,i]
            j += 1
        i += 1
        
    x = np.zeros(n)
    
    i = n-1
    while i >= 0:
        j = i
        x[i] = A1[i,n]
        while j < n-1:
            x[i] -= A1[i,j+1]*x[j+1]
            j += 1
        x[i] = x[i]/A1[i,i]
        i -= 1
    
    return x

In [18]:
def L2Norm(v):
    return la.norm(v,2)

In [19]:
b = np.array([1, 2, 3, 4])
db = np.array([0.1, 0, 0 , 0])
bp = b + db

x1 = GaussianElimination(A1,b)
x2 = GaussianElimination(A2,b)
xp1 = GaussianElimination(A1,bp)
xp2 = GaussianElimination(A2,bp)

print("Change in b = %.2f%%" % (100*L2Norm(bp-b)/L2Norm(b)))
print("Change in x1 = %.2f%%" % (100*L2Norm(xp1-x1)/L2Norm(x1)))
print("Change in x2 = %.2f%%" % (100*L2Norm(xp2-x2)/L2Norm(x2)))

Change in b = 1.83%
Change in x1 = 512.62%
Change in x2 = 10.48%


Which system of equations is more sensitive to small changes in $b$? How could you have guessed this from the answer to the Q3a?

### Matrix A1 is the most sensitive to small changes in b.
A change in b of 1.83% causes a change in x1 of 512.62%

I would have guessed this from Q3(a) because A1 is an ill-conditioned numbers which, in general, can be more sensitive to small changes in b.

Commit your notebook to git and push it to GitHub.

# Question 4: eigensystems

In this question we look at finding eigenvalues and eigenvectors

## Q4a

You may use any of NumPy's functions for this sub-question.

Find the eigenvalues and eigenvectors for the below matrix. Print out the eigenvalues sorted by descending value and explicitly show that eigenvalues/eigenvector pairs satisfy $Ax = \lambda x$.

In [20]:
A = np.array([[1,4,5],[4,5,4],[6,7,9]])

In [21]:
eigs = la.eig(A)

print("Eigenvalues = ", eigs[0])
print("Sorted Eigenvalues = ", np.sort(eigs[0]))
print("\nEigenvector 0 = ", eigs[1][:,0]) #Eigenvectors are collumns of eigs[1]
print("Eigenvector 1 = ", eigs[1][:,1])
print("Eigenvector 2 = ", eigs[1][:,2])

Eigenvalues =  [15.78569318 -2.05164645  1.26595327]
Sorted Eigenvalues =  [-2.05164645  1.26595327 15.78569318]

Eigenvector 0 =  [-0.39252008 -0.44424151 -0.80534308]
Eigenvector 1 =  [-0.89589878  0.36255198  0.25675168]
Eigenvector 2 =  [ 0.09818766 -0.77454142  0.6248558 ]


In [22]:
#Explicitly show A*x = EigenValue*x
print('Eigenvector 0')
print('Left Hand Side:',A @ eigs[1][:,0])
print('RightHand Side:',eigs[0][0]*eigs[1][:,0])

print('\nEigenvector 1')
print('Left Hand Side:',A @ eigs[1][:,1])
print('RightHand Side:',eigs[0][1]*eigs[1][:,1])

print('\nEigenvector 2')
print('Left Hand Side:',A @ eigs[1][:,2])
print('RightHand Side:',eigs[0][2]*eigs[1][:,2])

Eigenvector 0
Left Hand Side: [ -6.1962015   -7.01266016 -12.71289871]
RightHand Side: [ -6.1962015   -7.01266016 -12.71289871]

Eigenvector 1
Left Hand Side: [ 1.83806756 -0.74382849 -0.52676368]
RightHand Side: [ 1.83806756 -0.74382849 -0.52676368]

Eigenvector 2
Left Hand Side: [ 0.12430099 -0.98053324  0.79103825]
RightHand Side: [ 0.12430099 -0.98053324  0.79103825]


# Q4b

Without using NumPy's `la.eig()` (or similar functions), find the 4 eigenvalues and eigenvectors of the following matrix. Hint: Use the power method for the max eigenvalue (given below in `MaxEigenvalue`), the power method applied to the inverse for the smallest eigenvalue, and apply Hotelling's deflation twice to find the second and third eigenvalues/eigenvectors.

In [23]:
A = np.array([[2.65287619, 1.58615494, 1.47170961, 1.40586081],
       [1.58615494, 1.17577323, 0.72084593, 0.79382861],
       [1.47170961, 0.72084593, 1.18111969, 0.6654689 ],
       [1.40586081, 0.79382861, 0.6654689 , 0.84502394]])

In [24]:
def MaxEigenvalue(A, err):
    (m,n) = A.shape
    if(m != n):
        print("Matrix must be square")
        return
    
    # Create a random initial vector
    x = np.random.rand(m)
    
    lam = 0.1
    lamprev = 1
    while np.abs(1-lam/lamprev) > err:
        Ax      = A@x
        lamprev = lam
        lam = la.norm(Ax,2)/la.norm(x,2)
        x = Ax
        
    x = x/la.norm(x)
    return (lam,x)

In [25]:
# 4th, Max EigenValue/EigenVector 
EigenValue4 = MaxEigenvalue(A, 1e-15)[0]
EigenVector4 = MaxEigenvalue(A, 1e-15)[1]

In [26]:
# 3rd EigenVector
EigenValueB1 = EigenValue4
EigenVectorB1 = EigenVector4
B = A - EigenValueB1*np.outer(EigenVectorB1, EigenVectorB1)

EigenValue3 = MaxEigenvalue(B, 1e-15)[0]
EigenVector3 = MaxEigenvalue(B, 1e-15)[1]

In [27]:
# 2nd EigenVector
EigenValueB2 = EigenValue3
EigenVectorB2 = EigenVector3
C = B - EigenValueB2*np.outer(EigenVectorB2, EigenVectorB2)
EigenValue2 = MaxEigenvalue(C, 1e-15)[0]
EigenVector2 = MaxEigenvalue(C, 1e-15)[1]

In [28]:
# 1st EigenValue
# Test to see if A is invertable
if la.det(A) != 0: # A is Invertable
    EigenValue1 = 1/MaxEigenvalue(la.inv(A), 1e-15)[0]

In [29]:
# 1st EigenVector
# Can't get MinEigenvalueSymmetricPositiveDefinite() from notes to work properly, so take 3rd iteration of the Power Rule instead
EigenValueB1 = EigenValue2
EigenVectorB1 = EigenVector2
D = C - EigenValueB1*np.outer(EigenVectorB1, EigenVectorB1)
EigenVector1 = MaxEigenvalue(D, 1e-15)[1]

Check your answers against Numpy's results from `la.eig(A)`

In [30]:
EigenSystem = la.eig(A)
sortedEigenSystem = np.sort(EigenSystem[0])

#EigenValues
print("EigenValue1:\n Power Method: %.10f \n NumPy:        %.10f\n" %(EigenValue1,sortedEigenSystem[0]))
print("EigenValue2:\n Power Method: %.10f \n NumPy:        %.10f\n" %(EigenValue2,sortedEigenSystem[1]))
print("EigenValue3:\n Power Method: %.10f \n NumPy:        %.10f\n" %(EigenValue3,sortedEigenSystem[2]))
print("EigenValue4:\n Power Method: %.10f \n NumPy:        %.10f\n" %(EigenValue4,sortedEigenSystem[3]))

#EigenVectors
print("EigenVector1:", abs(EigenVector1))
print('NumPy        ', abs(EigenSystem[1][:,1]))

print("\nEigenVector2:", abs(EigenVector2))
print('NumPy        ', abs(EigenSystem[1][:,2]))

print("\nEigenVector3:", abs(EigenVector3))
print('NumPy        ', abs(EigenSystem[1][:,3]))

print("\nEigenVector4:", abs(EigenVector4))
print('NumPy        ', abs(EigenSystem[1][:,0]))

EigenValue1:
 Power Method: 0.0015297972 
 NumPy:        0.0015297972

EigenValue2:
 Power Method: 0.1986227827 
 NumPy:        0.1986227827

EigenValue3:
 Power Method: 0.4816692703 
 NumPy:        0.4816692703

EigenValue4:
 Power Method: 5.1729711998 
 NumPy:        5.1729711998

EigenVector1: [0.68301529 0.34802856 0.328043   0.55204531]
NumPy         [0.68301529 0.34802856 0.328043   0.55204531]

EigenVector2: [0.14168212 0.64758012 0.22152223 0.71518813]
NumPy         [0.14168212 0.64758011 0.22152224 0.71518814]

EigenVector3: [0.04054567 0.52331842 0.82477029 0.21035171]
NumPy         [0.04054567 0.52331842 0.82477029 0.21035171]

EigenVector4: [0.71538265 0.43087578 0.40381884 0.37350243]
NumPy         [0.71538265 0.43087578 0.40381884 0.37350243]


Commit your notebook to git and push it to GitHub. Upload your completed assignment to BrightSpace.